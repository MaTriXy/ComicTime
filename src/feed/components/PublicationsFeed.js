import React from "react";
import { GridList, GridTile } from "material-ui";
import { numberOfGridColumns } from "../../utils/windowUtils";
import LazyImage from "../../baseComponents/lazyImage/LazyImage";

const defaultCoverImageHeight = 350;
const defaultCoverImageWidth = 250;

class PublicationsFeed extends React.Component {
  constructor(props) {
    super(props);
    this.updateDimensions = this.updateDimensions.bind(this);
    this.sizeStatus = this.sizeStatus.bind(this);
  }

  componentDidMount() {
    window.addEventListener("resize", this.updateDimensions);
  }

  componentWillMount() {
    this.updateDimensions();
  }

  componentWillUnmount() {
    window.removeEventListener("resize", this.updateDimensions);
  }

  updateDimensions() {
    this.setState(this.sizeStatus());
  }

  sizeStatus() {
    return { width: window.innerWidth, height: window.innerHeight };
  }

  render() {
    const cellHeight = this.cellHeight();
    return (
      <GridList cols={this.numberOfGridColumns()} cellHeight={cellHeight}>
        {this.renderGridTiles(cellHeight)}
      </GridList>
    );
  }

  renderGridTiles(cellHeight) {
    return this.props.publications.map((publication, index) => (
      <GridTile
        key={index}
        title={publication.title}
        onClick={() => {
          this.props.onPublicationSelected(publication.path);
        }}
        subtitle={
          <span>
            <b>{publication.lastIssuesNumbers.join(", ")}</b>
            {" - "}
            {publication.releaseDate}
          </span>
        }
      >
        <LazyImage
          height={cellHeight}
          src={publication.image}
          placeholder={publication.image}
          alt={publication.title}
        />
      </GridTile>
    ));
  }

  numberOfGridColumns() {
    return numberOfGridColumns(this.state.width);
  }

  cellHeight() {
    const screenWidth = this.state.width;
    const columnWidth = screenWidth / this.numberOfGridColumns();
    return columnWidth * defaultCoverImageHeight / defaultCoverImageWidth;
  }
}

export default PublicationsFeed;
