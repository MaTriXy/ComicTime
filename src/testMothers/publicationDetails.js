import { IssueSummary, Publication } from "../publicationDetail/model";
import { Issue, IssuePage } from "../issueDetail/model";
import { Some } from "monet";

export const anyPublicationId = "the-flash-2016";
export const anyIssueId = "Annual1";

export const anyPublication = new Publication(
  "the-flash-2016",
  "https://comicimages.com/the-flash-1",
  "The Flash",
  Some("Ongoing"),
  "The best comic series in the world",
  Some("DC"),
  Some("January 2016"),
  Some("Peter Gomez"),
  [
    new IssueSummary("1", "the-flash-2016", "The Flash #1", "January 2016"),
    new IssueSummary("2", "the-flash-2016", "The Flash #2", "February 2016")
  ]
);

export const anyIssueSummary = new IssueSummary(
  "1",
  "detective-comic",
  "Detective Comic #1",
  "January 2018"
);

export const anyIssue = new Issue(
  "1",
  "detective-comic",
  "Detective Comic #1",
  2,
  [
    new IssuePage(1, "https://comicpages.com/1"),
    new IssuePage(2, "https://comicpages.com/2")
  ]
);
