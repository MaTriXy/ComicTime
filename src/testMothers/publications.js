import { Page, PublicationSummary } from "../feed/model";

export const anyListOfPublications = [
  new PublicationSummary(
    "uuid1",
    "http://readcomicsonline.ru/comic/deadpool-2017",
    "deadpool-2017",
    "deadpool (2017)",
    ["deadpool (2017) #1"],
    "http://readcomicsonline.ru/uploads/manga/deadpool-2017/cover/cover_250x350.jpg",
    "20/1/2017",
    "Marvel"
  ),
  new PublicationSummary(
    "uuid2",
    "http://readcomicsonline.ru/comic/volcanosaurus-2017",
    "volcanosaurus-2017",
    "Volcanosaurus (2017)",
    ["Volcanosaurus (2017) #2"],
    "http://readcomicsonline.ru/uploads/manga/volcanosaurus-2017/cover/cover_250x350.jpg",
    "04/5/2018",
    "American Mythology"
  )
];

export const anyFirstPageOfPublications = new Page(anyListOfPublications, 1, 1);
export const anySecondPageOfPublications = new Page(
  anyListOfPublications,
  2,
  2
);
