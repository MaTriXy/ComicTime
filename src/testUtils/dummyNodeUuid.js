jest.mock("node-uuid", () => {
  return {
    v4() {
      return "uuid";
    }
  };
});
