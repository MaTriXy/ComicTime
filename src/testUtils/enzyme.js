import React from "react";
import Enzyme from "enzyme";
import { createSerializer } from "enzyme-to-json";
import toJson from "enzyme-to-json";
import Adapter from "enzyme-adapter-react-16";
import { MuiThemeProvider } from "material-ui";
import comicTimeTheme from "../theme";

expect.addSnapshotSerializer(createSerializer({ mode: "deep" }));
Enzyme.configure({ adapter: new Adapter() });

export function mockHistory() {
  return {
    go: jest.fn(),
    goBack: jest.fn()
  };
}

export function translator(key) {
  return key;
}

export function mountWithMuyThemeProvider(component) {
  return Enzyme.mount(
    <MuiThemeProvider muiTheme={comicTimeTheme}>{component}</MuiThemeProvider>
  );
}

export function mountComponent(component) {
  return Enzyme.mount(component);
}

export function mountComponentAsJson(component) {
  return toJson(mountComponent(component));
}

export function shallowComponent(component) {
  return Enzyme.shallow(component);
}

export function shallowComponentAsJson(component) {
  return toJson(shallowComponent(component));
}

export function json(component) {
  return toJson(component);
}
