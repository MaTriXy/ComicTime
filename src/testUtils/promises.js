export async function flushPromises() {
  return new Promise(resolve => setImmediate(resolve));
}
