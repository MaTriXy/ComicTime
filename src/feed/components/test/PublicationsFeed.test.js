import React from "react";
import PublicationsFeed from "../PublicationsFeed";
import { shallowComponentAsJson } from "../../../testUtils/enzyme";
import { anyListOfPublications } from "../../../testMothers/publications";

describe("PublicationsFeed", () => {
  it("renders an empty publications feed", () => {
    const component = shallowComponentAsJson(
      <PublicationsFeed publications={[]} />
    );

    expect(component).toMatchSnapshot();
  });

  it("renders the publication feed based on the list of publications passed as props", () => {
    const component = shallowComponentAsJson(
      <PublicationsFeed publications={anyListOfPublications} />
    );

    expect(component).toMatchSnapshot();
  });
});
