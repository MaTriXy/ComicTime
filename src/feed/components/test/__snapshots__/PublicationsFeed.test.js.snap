// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PublicationsFeed renders an empty publications feed 1`] = `
<GridList
  cellHeight={358.4}
  cols={4}
  padding={4}
/>
`;

exports[`PublicationsFeed renders the publication feed based on the list of publications passed as props 1`] = `
<GridList
  cellHeight={358.4}
  cols={4}
  padding={4}
>
  <GridTile
    actionPosition="right"
    cols={1}
    containerElement="div"
    key="0"
    rows={1}
    subtitle={
      <span>
        <b>
          #1
        </b>
         - 
        20/1/2017
      </span>
    }
    title="deadpool (2017)"
    titleBackground="rgba(0, 0, 0, 0.4)"
    titlePosition="bottom"
  >
    <LazyImage
      alt="deadpool (2017)"
      height={358.4}
      placeholder="http://readcomicsonline.ru/uploads/manga/deadpool-2017/cover/cover_250x350.jpg"
      src="http://readcomicsonline.ru/uploads/manga/deadpool-2017/cover/cover_250x350.jpg"
    />
  </GridTile>
  <GridTile
    actionPosition="right"
    cols={1}
    containerElement="div"
    key="1"
    rows={1}
    subtitle={
      <span>
        <b>
          #2
        </b>
         - 
        04/5/2018
      </span>
    }
    title="Volcanosaurus (2017)"
    titleBackground="rgba(0, 0, 0, 0.4)"
    titlePosition="bottom"
  >
    <LazyImage
      alt="Volcanosaurus (2017)"
      height={358.4}
      placeholder="http://readcomicsonline.ru/uploads/manga/volcanosaurus-2017/cover/cover_250x350.jpg"
      src="http://readcomicsonline.ru/uploads/manga/volcanosaurus-2017/cover/cover_250x350.jpg"
    />
  </GridTile>
</GridList>
`;
